### PR TITLE
8339303: C2: dead node after failing to match cloned address expression

### PR DESCRIPTION
**Pull request by robcasloz**
This changeset prevents the x86 platform-specific logic from cloning address expressions consisting of two chained `AddP` nodes with a small constant offset each, such as in the following example:

example

Such patterns cannot be fully subsumed into x86 complex addressing modes, and cloning them can cause the matcher to introduce dead nodes that trigger a segmentation fault in the subsequent global code motion phase. See a detailed analysis of the failure in the [JBS issue description](https://bugs.openjdk.org/browse/JDK-8339303).

The changeset additionally extends the post-matching verification logic to check that no old node is reachable by travesing both node inputs and outputs. This extension would have caused the original test case to fail directly after matching with an informative assertion message rather than an opaque segmentation fault in an unrelated code generation phase.

Note that the pattern causing the failure should be in general optimized by `AddPNode::Ideal` into a single `AddP` node with the constant sum of the offsets. While [JDK-8343067](https://bugs.openjdk.org/browse/JDK-8343067) should address the missing optimization, this changeset proposes a complementary solution that is easily backportable and avoids relying on specific optimizations for correctness.

##### Functionality

- tier1-5 (linux-x64, windows-x64, macosx-x64, linux-aarch64, macosx-aarch64; release and debug mode).

##### Performance

- Tested performance on a set of standard benchmark suites (DaCapo, SPECjbb2015, SPECjvm2008). No significant change was observed.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8339303](https://bugs.openjdk.org/browse/JDK-8339303): C2: dead node after failing to match cloned address expression (**Bug** - P2)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) Review applies to [e85ba7cb](https://git.openjdk.org/jdk/pull/21829/files/e85ba7cbf919f6aa43a57a2a91793e70226ff24e)
 * [Vladimir Ivanov](https://openjdk.org/census#vlivanov) (@iwanowww - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21829/head:pull/21829` \
`$ git checkout pull/21829`

Update a local copy of the PR: \
`$ git checkout pull/21829` \
`$ git pull https://git.openjdk.org/jdk.git pull/21829/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21829`

View PR using the GUI difftool: \
`$ git pr show -t 21829`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21829.diff">https://git.openjdk.org/jdk/pull/21829.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21829#issuecomment-2454233961)
</details>

### CODE REVIEW

**Review by vnkozlov — APPROVED** [review]
Looks good. Yes, it looks like code expect LShift here instead of constant.

File: src/hotspot/share/opto/matcher.cpp [hunk 1]
```diff
@@ -170,17 +170,19 @@ void Matcher::verify_new_nodes_only(Node* xroot) {
...
       Node* in = n->in(j);
       if (in != nullptr) {
-        assert(C->node_arena()->contains(in), "dead node");
-        if (!visited.test(in->_idx)) {
-          worklist.push(in);
-        }
+        worklist.push(in);
       }
     }
+    for (uint j = 0; j < n->outcnt(); j++) {
```
Comment (iwanowww) [inline src/hotspot/share/opto/matcher.cpp:183]: Why don't you use DU iterator instead (e.g., `DUIterator_Fast`)?

**Review by iwanowww — APPROVED** [review]
Looks good.

**Review by iwanowww — APPROVED** [review]

**Comment by robcasloz** [comment]
Thanks for reviewing, @vnkozlov and @iwanowww!

**Pull request closed without merging**